File: src/memory/sibyl_bridge.py
#!/usr/bin/env python3
"""Sibyl Memory bridge — the ONLY place Drydock touches sibyl-memory-client.

sibyl-memory-client is a Python package (verified installed: 0.8.0). Drydock's
backend is Node/TypeScript. There is no Node SDK, so "in-process" (the
architecture.md §2.1 / memory finding "SDK, not MCP, for writes") is
satisfied the only way it can be here: this script imports MemoryClient
directly and calls its methods in-process *within this Python process* — it
does NOT go through the sibyl-memory-mcp server (a separate long-running
process with its own lossy 8-tool protocol, see §2.1). Node reaches this via
a single spawned subprocess per call (src/memory/sibylBridge.ts) — the
subprocess boundary is a Node<->Python language gap, not the MCP hop the
finding warned about; every method call below is a direct SDK call, full
field control (evaluated/acted/forward/extra/ts, set_reference, etc.).

Protocol: one JSON object on stdin, one JSON object on stdout.
  in:  {"op": "<method>", "args": {...}, "db_path"?: str, "tenant_id"?: str}
  out: {"ok": true, "result": <value>} | {"ok": false, "error": {"code","message","recovery"}}

Supported ops map 1:1 to MemoryClient methods used by architecture.md §2:
get_entity, set_entity, list_entities, write_event, read_events,
set_reference, get_reference, set_state, get_state.
"""
import json
import sys

from sibyl_memory_client import MemoryClient
from sibyl_memory_client.exceptions import NotFoundError, SibylMemoryError

DEFAULT_DB = "~/.sibyl-memory/memory.db"


def _client(db_path, tenant_id):
    # Free tier, no sibyl init, no bound account — the resolved decision in
    # architecture.md's banner and §9 Q5. tenant_id defaults to
    # MemoryClient's own DEFAULT_TENANT when not passed.
    kwargs = {"tier": "free"}
    if tenant_id:
        kwargs["tenant_id"] = tenant_id
    return MemoryClient.local(db_path or DEFAULT_DB, **kwargs)


def _dispatch(client, op, args):
    if op == "get_entity":
        try:
            return client.get_entity(args["category"], args["name"])
        except NotFoundError:
            # §6 pseudocode: "ledger = get_entity(...)  # may be NotFound" —
            # the Node caller treats this the same as "no ledger row yet".
            return None
    if op == "set_entity":
        return client.set_entity(
            args["category"], args["name"], args["body"], status=args.get("status")
        )
    if op == "list_entities":
        kwargs = {}
        if args.get("category") is not None:
            kwargs["category"] = args["category"]
        if args.get("status") is not None:
            kwargs["status"] = args["status"]
        if args.get("limit") is not None:
            kwargs["limit"] = args["limit"]
        return client.list_entities(**kwargs)
    if op == "write_event":
        return client.write_event(
            evaluated=args.get("evaluated"),
            acted=args.get("acted"),
            forward=args.get("forward"),
            extra=args.get("extra"),
            ts=args.get("ts"),
        )
    if op == "read_events":
        kwargs = {}
        if "limit" in args and args["limit"] is not None:
            kwargs["limit"] = args["limit"]
        if "since" in args and args["since"] is not None:
            kwargs["since"] = args["since"]
        if "until" in args and args["until"] is not None:
            kwargs["until"] = args["until"]
        return client.read_events(**kwargs)
    if op == "set_reference":
        client.set_reference(args["key"], args["body"], metadata=args.get("metadata"))
        return None
    if op == "get_reference":
        return client.get_reference(args["key"])
    if op == "set_state":
        client.set_state(args["key"], args["body"])
        return None
    if op == "get_state":
        return client.get_state(args["key"])
    raise ValueError(f"unknown op {op!r}")


def main() -> int:
    raw = sys.stdin.read()
    try:
        req = json.loads(raw)
        op = req["op"]
        args = req.get("args", {})
        client = _client(req.get("db_path"), req.get("tenant_id"))
        result = _dispatch(client, op, args)
        json.dump({"ok": True, "result": result}, sys.stdout)
        return 0
    except SibylMemoryError as e:
        json.dump(
            {
                "ok": False,
                "error": {
                    "code": getattr(e, "code", "SIBYL_MEMORY_ERROR"),
                    "message": str(e),
                    "recovery": getattr(e, "recovery", None),
                },
            },
            sys.stdout,
        )
        return 0
    except Exception as e:  # bridge/protocol errors — malformed request, etc.
        import traceback

        traceback.print_exc(file=sys.stderr)  # full trace to stderr for the Node caller
        json.dump(
            {"ok": False, "error": {"code": "BRIDGE_ERROR", "message": f"{type(e).__name__}: {e}"}},
            sys.stdout,
        )
        return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/memory/sibylBridge.ts
/**
 * Node <-> Python bridge for sibyl-memory-client (see sibyl_bridge.py header
 * for why a subprocess, not literally the same process, is the closest thing
 * to "in-process" available when the SDK is Python-only and Drydock's
 * backend is Node). One spawn per call — simple and crash-isolated; the
 * hackathon-scale incident volume this build expects doesn't need a
 * persistent worker.
 */
import { spawn } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const BRIDGE_SCRIPT = path.join(HERE, "sibyl_bridge.py");

// Override with SIBYL_PYTHON_BIN to point at the interpreter that has
// sibyl-memory-client installed. Defaults: on Windows, a verified absolute
// path for this dev machine (`python`/`python3` on PATH there can resolve to
// the Windows Store stub, which hangs on stdin instead of erroring); on
// Linux/macOS — e.g. the Railway container — plain `python3` on PATH.
const DEFAULT_PYTHON_BIN =
  process.platform === "win32"
    ? "C:\\Users\\Kelvin\\AppData\\Local\\Python\\bin\\python.exe"
    : "python3";
const PYTHON_BIN = process.env.SIBYL_PYTHON_BIN ?? DEFAULT_PYTHON_BIN;

const DB_PATH = process.env.SIBYL_MEMORY_DB; // undefined -> bridge's own default (~/.sibyl-memory/memory.db)
const TENANT_ID = process.env.SIBYL_TENANT_ID; // undefined -> DEFAULT_TENANT

export class SibylBridgeError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly recovery?: string | null,
  ) {
    super(message);
    this.name = "SibylBridgeError";
  }
}

interface BridgeResponse {
  ok: boolean;
  result?: unknown;
  error?: { code: string; message: string; recovery?: string | null };
}

function callBridge(op: string, args: Record<string, unknown>): Promise<unknown> {
  return new Promise((resolve, reject) => {
    // Node ALWAYS pipes the request as UTF-8. Without this, Python on Windows
    // opens stdin/stdout with the ANSI code page (cp1252) + surrogateescape —
    // so any non-ASCII byte in the payload is mis-decoded, and bytes that are
    // undefined in cp1252 (0x81/0x8D/0x8F/0x90/0x9D — all valid UTF-8
    // continuation bytes) become lone surrogates (\udc90 …) that
    // sibyl-memory-client then can't re-encode to UTF-8 for SQLite
    // ("UnicodeEncodeError: surrogates not allowed"). Pin both ends to UTF-8.
    const child = spawn(PYTHON_BIN, [BRIDGE_SCRIPT], {
      stdio: ["pipe", "pipe", "pipe"],
      env: { ...process.env, PYTHONIOENCODING: "utf-8", PYTHONUTF8: "1" },
    });
    let stdout = "";
    let stderr = "";
    child.stdout.setEncoding("utf8");
    child.stderr.setEncoding("utf8");
    child.stdout.on("data", (d) => (stdout += d));
    child.stderr.on("data", (d) => (stderr += d));
    child.on("error", (err) =>
      reject(new SibylBridgeError(`failed to spawn ${PYTHON_BIN}: ${err.message}`, "SPAWN_ERROR")),
    );
    child.on("close", (code) => {
      if (stdout.trim() === "") {
        reject(
          new SibylBridgeError(
            `sibyl_bridge.py produced no output (exit ${code}). stderr: ${stderr.slice(0, 2000)}`,
            "BRIDGE_NO_OUTPUT",
          ),
        );
        return;
      }
      let parsed: BridgeResponse;
      try {
        parsed = JSON.parse(stdout);
      } catch (err) {
        reject(
          new SibylBridgeError(
            `sibyl_bridge.py returned non-JSON: ${stdout.slice(0, 2000)}`,
            "BRIDGE_BAD_OUTPUT",
          ),
        );
        return;
      }
      if (!parsed.ok) {
        const e = parsed.error ?? { code: "UNKNOWN", message: "bridge reported failure with no error detail" };
        // Include the Python stderr (traceback) when present — a bare
        // "UnicodeEncodeError: …" message with no location is hard to debug.
        const detail = stderr.trim() ? `${e.message}\n--- bridge stderr ---\n${stderr.trim().slice(0, 4000)}` : e.message;
        reject(new SibylBridgeError(detail, e.code, e.recovery));
        return;
      }
      resolve(parsed.result);
    });
    child.stdin.write(JSON.stringify({ op, args, db_path: DB_PATH, tenant_id: TENANT_ID }), "utf8");
    child.stdin.end();
  });
}

// --- Typed wrappers, one per architecture.md §2 primitive ------------------

export function getEntity(category: string, name: string): Promise<{ body: unknown; status: string | null } | null> {
  return callBridge("get_entity", { category, name }) as Promise<any>;
}

export function listEntities(
  category?: string,
  opts: { status?: string; limit?: number } = {},
): Promise<Array<{ id: string; category: string; name: string; status: string | null; body: unknown }>> {
  return callBridge("list_entities", { category, ...opts }) as Promise<any>;
}

export function setEntity(
  category: string,
  name: string,
  body: Record<string, unknown>,
  status?: string,
): Promise<unknown> {
  return callBridge("set_entity", { category, name, body, status });
}

export function writeEvent(event: {
  evaluated?: unknown;
  acted?: unknown;
  forward?: unknown;
  extra?: unknown;
  ts?: string;
}): Promise<string> {
  return callBridge("write_event", event) as Promise<string>;
}

export function readEvents(opts: { limit?: number; since?: string; until?: string } = {}): Promise<any[]> {
  return callBridge("read_events", opts) as Promise<any[]>;
}

export function setReference(
  key: string,
  body: unknown,
  metadata?: Record<string, unknown>,
): Promise<void> {
  return callBridge("set_reference", { key, body, metadata }) as Promise<void>;
}

export function getReference(key: string): Promise<{ body: unknown; metadata: unknown; updated_at: string } | null> {
  return callBridge("get_reference", { key }) as Promise<any>;
}

export function setState(key: string, body: unknown): Promise<void> {
  return callBridge("set_state", { key, body }) as Promise<void>;
}

export function getState(key: string): Promise<{ body: unknown; updated_at: string } | null> {
  return callBridge("get_state", { key }) as Promise<any>;
}
